grudge/trace_pair.py

```diff
@@ -272,61 +277,130 @@ def connected_ranks(dcoll: DiscretizationCollection):
     return get_connected_partitions(dcoll._volume_discr.mesh)
 
 
+def _flatten_to_numpy_from_container(
+        actx: ArrayContext, ary: ArrayOrContainerT) -> ArrayT:
+    """Flattens an arbitrary array container *ary* into a one-dimensional
+    numpy array suitable for communication via mpi4py.
+    """
+    if is_array_container(ary):
+        flat_data = []
+        # NOTE: The ordering of data in the flattened array will be determined
+        # by the ordering of key/value pairs returned by *serialize_container*.
+        # This is important to remember because the function
+        # *_unflatten_from_numpy_to_container* relies on this same ordering
+        # to unflatten the concatenated array.
+        for _, value in serialize_container(to_numpy(ary, actx)):
+            # NOTE: *to_numpy* can return numpy object arrays whose components
+            # are numpy arrays. As a result, we need to flatten these as well.
+            if isinstance(value, np.ndarray) and value.dtype == "O":
+                flat_data.append(np.concatenate(value))
+            else:
+                flat_data.append(value)
+        # All data is concatenated into a single numpy array
+        flattened_np_ary = np.concatenate(flat_data)
```

Comment:
For a deeply nested structure, this does a lot of recursive concatenating, amounting to a lot of copying. Instead, build a list of flattened arrays and concatenate them all at once.